### PR TITLE
#34146 - differential attention implementation for bert

### LINKED ISSUE #34146: Differential Attention implementation for BERT.
### Feature request

The paper "Differential Transformers" implements a differential attention mechanism which calculates the attention scores as the difference between two separate softmax attention maps leading to better long-context modeling and key information retrieval.

LINK : [Paper](https://arxiv.org/pdf/2410.05258)

### Motivation

Although the paper focuses on decoder only models, I think this differential attention mechanism might be helpful with encoder only models as well.

### Your contribution

I can work on this feature, if the HuggingFace team considers it as a valuable addition.

### PR DESCRIPTION
**Pull request by Chirayu-Tripathi**
# What does this PR do?
Fixes https://github.com/huggingface/transformers/issues/34146
Implements differential attention for BERT

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! Do you know if there were any official checkpoints trained with this attention? 
In general we would add a new model for DifferentialBert as it does not comput the same thing as eager, flash or sdpa

**Comment by Chirayu-Tripathi** [comment]
> Hey! Do you know if there were any official checkpoints trained with this attention? In general we would add a new model for DifferentialBert as it does not comput the same thing as eager, flash or sdpa

Hey @ArthurZucker,
Thanks for the review, but I am afraid they did not release the weights, I am hoping they would release it with time, or we might need to pre-train it.

**Comment by Chirayu-Tripathi** [comment]
Hey @ArthurZucker,
Sorry to bother you, I was wondering if we can still merge this, because I think the code can still be used to pre-train the model, or   does it require some major changes?

**Comment by ArthurZucker** [comment]
Hey! Let's just keep this as is until we have released checkpoints! I would advise you to share your code on the hub this way people can use it! 🤗 https://huggingface.co/docs/transformers/v4.34.0/en/custom_models for more details!
